api/src/main/java/org/eclipse/microprofile/config/spi/ConfigValue.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.eclipse.microprofile.config.spi;
+
+import java.util.Objects;
+
+/**
+ * The ConfigValue is a metadata object that holds additional information after the lookup of a configuration.
+ * <p>
+ *
+ * Right now, it is able to hold information like the configuration name, value, the Config Source from where
+ * the configuration was loaded, the ordinal of the Config Source and a line number from where the configuration was
+ * read if exists.
+ * <p>
+ *
+ * This is used together with {@link ConfigValueConfigSource} and {@link ConfigSourceInterceptor} to expose the
+ * Configuration lookup metadata.
+ *
+ * @author <a href="mailto:radcortez@yahoo.com">Roberto Cortez</a>
+ */
+public class ConfigValue {
+    private final String name;
+    private final String value;
+    private final String configSourceName;
+    private final int configSourceOrdinal;
+
+    private final int lineNumber;
+
+    private ConfigValue(final ConfigValueBuilder builder) {
+        this.name = builder.name;
+        this.value = builder.value;
+        this.configSourceName = builder.configSourceName;
+        this.configSourceOrdinal = builder.configSourceOrdinal;
+        this.lineNumber = builder.lineNumber;
```

Comment:
I would remove this as the config could be in a database and the linenumber is not applicable. If it is in a file, string search is good enough.

Reply:
I would not remove this.  If the backend has a line number, this can return it.  If the backend does not, it can return a sentinel (such as -1) indicating that is the case.  Having the line number allows useful diagnostics to be printed, without requiring the larger overhead of having special location classes.

This is similar to what is done in libraries such as JAXP.

Reply:
We implemented this in SmallRye Config and I find it very useful. Its kinda like having the line number in the stacktrace :)

Reply:
why would the line number matter at all for the end user use case? Any developer would just look at the file. Also note that in yaml, json or xml you might not even be able to define a line number for the resulting key because the value is a result of information from many lines together.

Reply:
Other libraries do report the line number like JAXP, which is useful for debug diagnosis. Yes, you can look into the file, but again I think it is a nice addition. Anyway, this has caused a lot of discussion back and forward, so I'll most likely remove it.

Also, I believe the reason why we are adding a new method to `ConfigSource` has nothing to do with the `lineNumber`. In my opinion, the `ConfigSource` should have a way to return an object view with metadata information about the lookup. Right now all the information can be acquired by existing methods, but that doesn't mean that additional information may not be included in the future.